packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts

```diff
@@ -173,6 +173,50 @@ export interface UserPoolClientOptions {
    * @default - see defaults in `OAuthSettings`
    */
   readonly oAuth?: OAuthSettings;
+
+  /**
+   * The attributes this client will be able to read.
+   *
+   * This should be used to restrict the attributes the client will be
+   * able to read.
+   *
+   * @default undefined all attributes are readable
+   */
+  readonly readAttributes?: UserPoolClientAttributes;
+
+  /**
+   * The attributes this client will be able to write.
+   *
+   * This should be used to restrict the attributes the client will be able to write,
+   * e.g. the client used by a webapp should not be able to set the `customerPlan` attribute,
+   * while the client used by the billing backend should.
+   *
+   * @default undefined all attributes are writable
+   */
+  readonly writeAttributes?: UserPoolClientAttributes;
+}
+
+/**
+ * Represent a list of attributes (standard and custom)
+ */
+export interface UserPoolClientAttributes {
+  /**
+   * A list of standard attributes
+   *
+   * @default []
+   */
+  readonly standard?: StandardAttribute[]
```

Comment:
`StandardAttribute` is currently not exported. Instead of using this, can we reuse `RequiredAttributes` here?

Reply:
I looked into the `RequiredAttributes` enum but the `email_verified` and `phone_verified` attributes are missing.
Do you think it would be better to refactor `StandardAttribute` to be an interface extending `RequiredAttributes`?

```ts
export interface StandardAttributes extends RequiredAttributes {
  emailVerified?: boolean
  phoneVerified?: boolean
  allStandardAttributes?: boolean // this could be added to make easier to flag all attributes
}
```

What do you think? Should I refactor to the `interface`?

Reply:
Extending `RequiredAttributes` and adding `emailVerified` and `phoneVerified` to it sounds like a good idea. However, it can't be named `StandardAttributes` because they're not under [the list of standard attributes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#cognito-user-pools-standard-attributes). How about using a different name for this, such as, `UserPoolClientAttributes`?

> `allStandardAttributes?: boolean // this could be added to make easier to flag all attributes`

As for this, we could do create a method along the lines of `UserPoolClientAttributes.allStandardAttributes()`.

Take a look at how we've done something similar with `AccessLog` - https://github.com/aws/aws-cdk/blob/40fa93a22ffbdf18b0563d1cef63bbf5814dcc3f/packages/%40aws-cdk/aws-apigateway/lib/access-log.ts#L484-L519

Reply:
I'm sorry for the delay.

What do you think about creating an interface to hold all attributes to be set for a client (both standard and custom)
```ts
export interface IUserPoolClientAttributes extends RequiredAttributes {
  readonly emailVerified?: boolean;
  readonly phoneVerified?: boolean;
  // this will allow to set custom and standard attributes from this interface
  [customAttribute: string]: boolean | undefined;
}
```

and a class for conveniently create instances of that interface
```ts
export class UserPoolClientAttributes {
  public static allStandardAttributes(additionalCustomAttributes: string[] = []): IUserPoolClientAttributes {
    return {
      email: true,
      phone: true,
      // ... all other standard attributes
      custom:attribute1: true,
      custom:attribute2: true,
      // any other custom attribute specified in `additionalCustomAttributes`
    }
  }
}
```

The usage will be:
```ts
const readAttributes = UserPoolClientAttributes.allStandard(['attr1', 'attr2']);
const writeAttributes = UserPoolClientAttributes.allStandard();
const pool = new UserPool(this, 'Pool');
pool.addClient('webapp-client', {
  readAttributes,
  writeAttributes,
});
```

One concern about extending `RequiredAttributes` is that its docstrings are not appropriate for the use in app client. (e.g. "Whether the user's postal address is a required attribute.") 
https://github.com/aws/aws-cdk/blob/40fa93a22ffbdf18b0563d1cef63bbf5814dcc3f/packages/%40aws-cdk/aws-cognito/lib/user-pool-attr.ts#L1-L23

Reply:
> One concern about extending RequiredAttributes is that its docstrings are not appropriate for the use in app client.

We could modify the doc strings so that they are a bit more generic such as 'the user's postal address'. 

For the rest of your code, it's quite difficult to discuss this and comment on your code. Do you mind making these changes and then I can comment on whether they fit?
The important thing to consider is the user experience when a user uses this library. It needs to be nice and slick.